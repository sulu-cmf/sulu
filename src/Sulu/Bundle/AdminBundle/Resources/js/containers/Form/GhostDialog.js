// @flow
import React from 'react';
import {action, observable} from 'mobx';
import {observer} from 'mobx-react';
import Dialog from '../../components/Dialog';
import SingleSelect from '../../components/SingleSelect';
import {translate} from '../../utils/Translator';
import ghostDialogStyles from './ghostDialog.scss';

type Props = {
    locales: Array<string>,
    onCancel: () => void,
    onConfirm: (locale: string) => void,
    open: boolean,
};

@observer
export default class GhostDialog extends React.Component<Props> {
    @observable selectedLocale: string;

    constructor(props: Props) {
        super(props);

        this.selectedLocale = this.props.locales[0];
    }

    handleCancel = () => {
        this.props.onCancel();
    };

    handleConfirm = () => {
        this.props.onConfirm(this.selectedLocale);
    };

    @action handleLocaleChange = (locale: string | number) => {
        if (typeof locale !== 'string') {
            throw new Error('Only strings are accepted as locales! This should not happen and is likely a bug.');
        }

        this.selectedLocale = locale;
    };

    render() {
        const {
            locales,
            open,
        } = this.props;

        return (
            <Dialog
                cancelText={translate('sulu_admin.no')}
                onCancel={this.handleCancel}
                confirmText={translate('sulu_admin.yes')}
                onConfirm={this.handleConfirm}
                open={open}
                title={translate('sulu_admin.ghost_dialog_title')}
            >
                <div className={ghostDialogStyles.ghostDialog}>
                    <p>{translate('sulu_admin.ghost_dialog_description')}</p>
                    <label className={ghostDialogStyles.label}>{translate('sulu_admin.choose_language')}</label>
                    <div className={ghostDialogStyles.localeSelect}>
                        <SingleSelect onChange={this.handleLocaleChange} value={this.selectedLocale}>
                            {locales.map((locale) => (
                                <SingleSelect.Option key={locale} value={locale}>
                                    {locale}
                                </SingleSelect.Option>
                            ))}
                        </SingleSelect>
                    </div>
                </div>
            </Dialog>
        );
    }
}
