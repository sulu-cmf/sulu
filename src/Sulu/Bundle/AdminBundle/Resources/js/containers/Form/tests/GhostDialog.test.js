// @flow
import React from 'react';
import {mount} from 'enzyme';
import pretty from 'pretty';
import GhostDialog from '../GhostDialog';

jest.mock('../../../utils/Translator', () => ({
    translate: jest.fn(),
}));

afterEach(() => {
    if (document.body) {
        document.body.innerHTML = '';
    }
});

test('Should render a Dialog', () => {
    const body = document.body;

    mount(<GhostDialog locales={['en', 'de']} onCancel={jest.fn()} onConfirm={jest.fn()} open={true} />);

    expect(pretty(body ? body.innerHTML : '')).toMatchSnapshot();
});

test('Should call onCancel callback if user chooses not to copy content', () => {
    const cancelSpy = jest.fn();
    const ghostDialog = mount(
        <GhostDialog locales={['en', 'de']} onCancel={cancelSpy} onConfirm={jest.fn()} open={true} />
    );

    ghostDialog.find('Button[skin="secondary"]').simulate('click');

    expect(cancelSpy).toBeCalledWith();
});

test('Should call onConfirm callback with chosen locale if user chooses to copy content', () => {
    const confirmSpy = jest.fn();
    const ghostDialog = mount(
        <GhostDialog locales={['en', 'de']} onCancel={jest.fn()} onConfirm={confirmSpy} open={true} />
    );

    ghostDialog.find('SingleSelect').prop('onChange')('de');
    ghostDialog.find('Button[skin="primary"]').simulate('click');

    expect(confirmSpy).toBeCalledWith('de');
});
