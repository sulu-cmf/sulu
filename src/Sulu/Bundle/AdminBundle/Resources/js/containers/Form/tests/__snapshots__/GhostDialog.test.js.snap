// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render a Dialog 1`] = `
"<div>
  <div class=\\"backdrop visible fixed\\"></div>
</div>
<div>
  <div class=\\"dialogContainer\\">
    <div class=\\"dialog\\">
      <section class=\\"content\\">
        <header></header>
        <article>
          <div class=\\"ghostDialog\\">
            <p></p><label class=\\"label\\"></label>
            <div class=\\"localeSelect\\">
              <div class=\\"select\\"><button class=\\"displayValue default\\" type=\\"button\\">
                  <div aria-label=\\"en\\" class=\\"croppedText\\" title=\\"en\\">
                    <div aria-hidden=\\"true\\" class=\\"front\\">e</div>
                    <div aria-hidden=\\"true\\" class=\\"back\\"><span>n</span></div>
                    <div class=\\"whole\\">en</div>
                  </div><span aria-label=\\"su-angle-down\\" class=\\"toggle su-angle-down\\"></span>
                </button></div>
            </div>
          </div>
        </article>
        <footer><button class=\\"button secondary\\" type=\\"button\\"><span class=\\"text\\"></span></button><button class=\\"button primary\\" type=\\"button\\"><span class=\\"text\\"></span></button></footer>
      </section>
    </div>
  </div>
</div>"
`;
